lib/internal/modules/esm/hooks.js

```diff
@@ -121,6 +122,9 @@ class Hooks {
     ],
   };
 
+  #lastInstanceId = 0;
+  #loaderInstances = [];
```

Comment:
I think we should not use a simple number like this for `id`. Instead, we should use a `Map` and some string-friendly id (because maps are **significantly** more efficient with string-based indexes than numeric that gets cast back and forth). It also makes mutations far more efficient, as well as being inherently better on memory (because an array must be pre-allocated space, whereas a Map takes up only exactly the amount of space it needs).

```suggestion
  #customLoaders = new SafeMap();
```

Reply:
We need to store the order of the loaders, a `Map` doesn't seem adequate. Also, it's the first time I hear that `Map` would be more memory efficient than arrays, I'd be very surprised if that was the case.

Reply:
Maps are ordered, but I agree

Reply:
Yes, the sequence of Map items is guaranteed (insertion order): [MDN: Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)

> The Map object holds key-value pairs and remembers the original insertion order of the keys.

---

> Also, it's the first time I hear that Map would be more memory efficient than arrays, I'd be very surprised if that was the case.

Maps are better than Arrays in terms of both space and time efficiency:

Arrays are pre-sized. V8 pre-allocates a rather significant size¹ for an array in order to avoid having to move it (because the entire array is stored sequentially in memory) which is a very expensive operation.

Maps are _not_ stored sequentially in memory, so there is no "move" operation when they grow.

¹ unless a size is specifically provided like `new Array(2)` (then the array will be exactly sized); however this can be a de-op do to data type of the items themselves (under the hood, V8 can use a better-suited type of array, but it doesn't do that optimisation when you strong-arm it, so unless the items in the array are all the same data-type, generally a primitive, it's better to not force a size).

Sets are similarly better for space and time efficiency (and are also not stored sequentially in memory).

Reply:
> Also worth noting: Maps and Sets are _significantly_ better for rapid mutation than Arrays, and that's exactly what we expect to happen here.

I don't think we expect that, I would expect that `unregister` to not be useful to most applications – I've never seen anyone request it, but maybe there are some use cases I'm not seeing.

Reply:
> Sets are similarly better for space and time efficiency (and are also not stored sequentially in memory).
> 
> Also worth noting: Maps and Sets are _significantly_ better for rapid mutation than Arrays, and that's exactly what we expect to happen here.

How did you come to that conclusion? Is there a way to check if this is true?

From my understanding, and with some pointers from @targos, I don't see how this could be true. 

<details>

In V8, `Set` and `Map` instances have a hash table associated with them which seems to be stored sequentially in memory:

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/builtins/builtins-collections-gen.cc#L1736
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/builtins/builtins-collections-gen.cc#L1789-L1791

https://github.com/nodejs/node/blob/edd537ca2f38b94738fe25c2dc9b8c21bc7847f2/deps/v8/src/runtime/runtime-collections.cc#L20-L26

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/objects/ordered-hash-table.cc#L69-L93

From the above, I understand the `Set` or `Map` hash table capacity is doubled everytime it reaches its capacity, so not so memory efficient. Now that `Rehash` function has to recompute the hash for each elements for the new table:

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/objects/ordered-hash-table.cc#L259-L261
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/objects/ordered-hash-table.cc#L278-L298

In comparison, `%Array.prototype.push%` has a fast path, and seems also simpler (because no hashing involved), and also has optimizations based on what type of object you add to the array. The object path:

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/builtins/builtins-array-gen.cc#L340
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/builtins/builtins-array-gen.cc#L362-L374
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/builtins/builtins-array-gen.cc#L405-L410

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.cc#L3446-L3450
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.cc#L3458-L3463
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.cc#L3429-L3444

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.cc#L5594-L5598
https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.cc#L5612-L5620

So if the preallocated space is not enough, the algorithm to copy the array elements is a bit tedious, so I'll stop there: https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.cc#L5249-L5255

</details>

It seems to me that when you are storing objects, they won't be copied over wether you use an array or a `Set` (I assume the same is true for object/`Map`), the engine will only store a reference (either in its hash table, or straight in the memory). In both cases, the engine pre-allocates some space, and when it's not enough, has to find another memory address where it can fit, and has to copy data over. In addition, for `Map` and `Set`, it has to recalculate the hashes.

TL;DR, `Set` is not more memory nor performant efficient when it comes it changing data size, it's more efficient to ensure uniqueness and `Set.prototype.has` is much more efficient than `Array.prototype.includes`, otherwise you're better off with an array.

Reply:
> How did you come to that conclusion? Is there a way to check if this is true?

A couple years ago I did a benchmark hammering each with adds & deletes, and the results were like 50:1 Map vs plain object and Map vs Array, and like 20–10:1 Set vs Array (presumably Set was less good than Map because of the overhead enforcing uniqueness?); the performance gap widened non-linearly as size increased (mutations to the Map remained about the same cost, but mutations to Array became more and more expensive as the content within the Array got bigger).

I don't imagine that since then Map and Set got _less_ performant or for Arrays to have changed much.

> TL;DR, `Set` is not more memory nor performant efficient when it comes it changing data

I can't find the article anymore; it was written by one of the V8 people (or maybe Jake Archibald?) and specifically talked about how Map and Set are stored non-sequentially. Based on your investigation, it sounds like Map and Set internal hashtable (that remember meta information) is stored sequentially (which makes sense), but that the items are not necessarily (because the hashtable contains references/pointers, which can point anywhere). Moving a hashtable would be significantly less costly than moving a huge array with huge data.

But I think we can ignore Array vs Set for this PR because we want direct access, so it would be Array vs Map.